modules/core/src/core/animation-loop-proxy.js

```diff
@@ -106,12 +139,49 @@ export default class AnimationLoopProxy {
       this._stopped = true;
       this.props.onFinalize(this);
     }
-    this.props.worker.postMessage({command: 'stop'});
+    this.worker.postMessage({command: 'stop'});
     return this;
   }
 
   // PRIVATE METHODS
 
+  _onMessage(evt) {
+    switch (evt.data.command) {
+      case 'addEventListener':
+        this.canvas.addEventListener(evt.data.type, this._onEvent);
+        break;
+
+      case 'removeEventListener':
+        this.canvas.removeEventListener(evt.data.type, this._onEvent);
+        break;
+
+      default:
+    }
+  }
+
+  _onEvent(evt) {
+    const devicePixelRatio = this.props.useDevicePixels ? (window.devicePixelRatio || 1) : 1;
```

Comment:
Is this code common to on-screen and off-screen? If it is only for offscreen, isn't `useDevicePixels` always false?

Reply:
`useDevicePixels` is false on the worker thread but defined by user on the main thread.